File: machine/cpp/machine/devices/vdp/vram.cpp
#include "machine/devices/vdp/vram.h"
#include "machine/devices/vdp/rpu.h"
#include "machine/memory/map.h"
#include <algorithm>
#include <string>
#include <utility>

namespace bmsx {
namespace {

inline u32 vramSurfaceByteSize(u32 width, u32 height) {
	return width * height * 4u;
}

} // namespace

VdpVramUnit::VdpVramUnit(VdpFrameBufferSize frameBufferSize, VdpEntropySeeds entropySeeds)
	: rpuVram(static_cast<size_t>(VRAM_STAGING_SIZE) + static_cast<size_t>(VRAM_TEXTURE_SIZE))
	, rpuVramPageRevisions(VDP_RPU_PARAM_MEM_PAGE_COUNT)
	, m_garbageScratch(VRAM_GARBAGE_CHUNK_BYTES)
	, m_configuredFrameBufferSize(frameBufferSize)
	, m_machineSeed(entropySeeds.machineSeed)
	, m_bootSeed(entropySeeds.bootSeed) {
	configureFrameBufferSurface(m_configuredFrameBufferSize.width, m_configuredFrameBufferSize.height);
}

void VdpVramUnit::initializeFrameBuffer() {
	VramGarbageStream stream{m_machineSeed, m_bootSeed, VRAM_GARBAGE_SPACE_SALT, VRAM_STAGING_BASE};
	fillVramGarbageScratch(rpuVram.data(), rpuVram.size(), stream);
	bumpVdpRpuVramPageRevisions(rpuVramPageRevisions.data(), 0u, rpuVram.size());
	configureFrameBufferSurface(m_configuredFrameBufferSize.width, m_configuredFrameBufferSize.height);
}

bool VdpVramUnit::writeRpuVram(u32 addr, const u8* bytes, size_t srcOffset, size_t length) {
	if (addr < VRAM_STAGING_BASE) {
		return false;
	}
	const size_t offset = static_cast<size_t>(addr - VRAM_STAGING_BASE);
	if (offset > rpuVram.size() || length > rpuVram.size() - offset) {
		return false;
	}
	for (size_t index = 0u; index < length; ++index) {
		rpuVram[offset + index] = bytes[srcOffset + index];
	}
	bumpVdpRpuVramPageRevisions(rpuVramPageRevisions.data(), static_cast<u32>(offset), length);
	return true;
}

bool VdpVramUnit::readRpuVram(u32 addr, u8* out, size_t length) const {
	if (addr < VRAM_STAGING_BASE) {
		return false;
	}
	const size_t offset = static_cast<size_t>(addr - VRAM_STAGING_BASE);
	if (offset > rpuVram.size() || length > rpuVram.size() - offset) {
		return false;
	}
	for (size_t index = 0u; index < length; ++index) {
		out[index] = rpuVram[offset + index];
	}
	return true;
}

// start repeated-sequence-acceptable -- VRAM row streaming keeps read/write loops direct; callback helpers would add hot-path overhead.
void VdpVramUnit::writeSurfaceBytes(VdpSurfaceBacking& surface, u32 offset, const u8* bytes, size_t srcOffset, size_t length) {
	const u32 stride = surface.surfaceWidth * 4u;
	size_t remaining = length;
	size_t cursor = srcOffset;
	u32 row = offset / stride;
	u32 rowOffset = offset - row * stride;
	while (remaining > 0u) {
		const u32 rowAvailable = stride - rowOffset;
		const u32 rowBytes = static_cast<u32>(std::min<size_t>(remaining, rowAvailable));
		updateCpuReadback(surface, bytes, cursor, rowBytes, rowOffset / 4u, row);
		remaining -= rowBytes;
		cursor += rowBytes;
		row += 1u;
		rowOffset = 0u;
	}
}

void VdpVramUnit::readSurfaceBytes(const VdpSurfaceBacking& surface, u32 offset, u8* out, size_t length) const {
	const u32 stride = surface.surfaceWidth * 4u;
	size_t remaining = length;
	size_t cursor = 0u;
	u32 row = offset / stride;
	u32 rowOffset = offset - row * stride;
	while (remaining > 0u) {
		const u32 rowAvailable = stride - rowOffset;
		const u32 rowBytes = static_cast<u32>(std::min<size_t>(remaining, rowAvailable));
		const size_t srcOffset = static_cast<size_t>(row) * static_cast<size_t>(stride) + static_cast<size_t>(rowOffset);
		for (u32 byteIndex = 0u; byteIndex < rowBytes; ++byteIndex) {
			out[cursor + byteIndex] = surface.cpuReadback[srcOffset + byteIndex];
		}
		remaining -= rowBytes;
		cursor += rowBytes;
		row += 1u;
		rowOffset = 0u;
	}
}
// end repeated-sequence-acceptable

void VdpVramUnit::setSurfaceLogicalDimensions(VdpSurfaceBacking& surface, u32 width, u32 height) {
	const u32 size = vramSurfaceByteSize(width, height);
	if (surface.surfaceWidth == width && surface.surfaceHeight == height) {
		return;
	}
	std::vector<u8> previous;
	previous.swap(surface.cpuReadback);
	surface.surfaceWidth = width;
	surface.surfaceHeight = height;
	surface.cpuReadback.resize(static_cast<size_t>(size));
	seedSurfacePixels(surface);
	const size_t copyBytes = previous.size() < surface.cpuReadback.size() ? previous.size() : surface.cpuReadback.size();
	for (size_t index = 0u; index < copyBytes; ++index) {
		surface.cpuReadback[index] = previous[index];
	}
}

bool VdpVramUnit::frameBufferContains(u32 addr, size_t length) const {
	const VdpSurfaceBacking& surface = m_frameBufferSurface;
	return addr >= surface.baseAddr && addr + length <= surface.baseAddr + surface.capacity;
}

VdpVramState VdpVramUnit::captureState() const {
	VdpVramState state;
	state.rpuVram = rpuVram;
	state.surfacePixels = captureSurfacePixels();
	return state;
}

void VdpVramUnit::restoreState(const VdpVramState& state) {
	for (size_t index = 0u; index < state.rpuVram.size(); ++index) {
		rpuVram[index] = state.rpuVram[index];
	}
	bumpVdpRpuVramPageRevisions(rpuVramPageRevisions.data(), 0u, state.rpuVram.size());
	for (const VdpSurfacePixelsState& surface : state.surfacePixels) {
		restoreSurfacePixels(surface);
	}
}

u32 VdpVramUnit::trackedUsedBytes() const {
	return static_cast<u32>(rpuVram.size()) + static_cast<u32>(vramSurfaceByteSize(m_frameBufferSurface.surfaceWidth, m_frameBufferSurface.surfaceHeight));
}

u32 VdpVramUnit::trackedTotalBytes() const {
	return VRAM_STAGING_SIZE + VRAM_TEXTURE_SIZE + VRAM_FRAMEBUFFER_SIZE;
}

void VdpVramUnit::configureFrameBufferSurface(u32 width, u32 height) {
	const u32 size = vramSurfaceByteSize(width, height);
	m_frameBufferSurface.baseAddr = VRAM_FRAMEBUFFER_BASE;
	m_frameBufferSurface.capacity = VRAM_FRAMEBUFFER_SIZE;
	m_frameBufferSurface.surfaceId = VDP_RD_SURFACE_FRAMEBUFFER;
	m_frameBufferSurface.surfaceWidth = width;
	m_frameBufferSurface.surfaceHeight = height;
	m_frameBufferSurface.cpuReadback.resize(static_cast<size_t>(size));
	seedSurfacePixels(m_frameBufferSurface);
}

std::vector<VdpSurfacePixelsState> VdpVramUnit::captureSurfacePixels() const {
	std::vector<VdpSurfacePixelsState> surfaces;
	surfaces.reserve(1u);
	VdpSurfacePixelsState state;
	state.surfaceId = m_frameBufferSurface.surfaceId;
	state.surfaceWidth = m_frameBufferSurface.surfaceWidth;
	state.surfaceHeight = m_frameBufferSurface.surfaceHeight;
	state.pixels = m_frameBufferSurface.cpuReadback;
	surfaces.push_back(std::move(state));
	return surfaces;
}

void VdpVramUnit::restoreSurfacePixels(const VdpSurfacePixelsState& state) {
	m_frameBufferSurface.cpuReadback = state.pixels;
}

void VdpVramUnit::updateCpuReadback(VdpSurfaceBacking& surface, const u8* bytes, size_t srcOffset, size_t length, u32 x, u32 y) {
	const u32 stride = surface.surfaceWidth * 4u;
	const size_t offset = static_cast<size_t>(y) * static_cast<size_t>(stride) + static_cast<size_t>(x) * 4u;
	for (size_t index = 0u; index < length; ++index) {
		surface.cpuReadback[offset + index] = bytes[srcOffset + index];
	}
}

void VdpVramUnit::seedSurfacePixels(VdpSurfaceBacking& surface) {
	const size_t rowPixels = static_cast<size_t>(surface.surfaceWidth);
	const size_t maxPixels = m_garbageScratch.size() / 4u;
	surface.cpuReadback.resize(vramSurfaceByteSize(surface.surfaceWidth, surface.surfaceHeight));
	VramGarbageStream stream{m_machineSeed, m_bootSeed, VRAM_GARBAGE_SPACE_SALT, surface.baseAddr};
	const size_t rowBytes = rowPixels * 4u;
	const u32 height = surface.surfaceHeight;
	if (rowBytes <= m_garbageScratch.size()) {
		const size_t rowsPerChunk = std::max<size_t>(1u, m_garbageScratch.size() / rowBytes);
		for (u32 y = 0u; y < height; ) {
			const size_t rows = std::min<size_t>(rowsPerChunk, height - y);
			const size_t chunkBytes = rowBytes * rows;
			fillVramGarbageScratch(m_garbageScratch.data(), chunkBytes, stream);
			for (size_t row = 0u; row < rows; ++row) {
				const size_t rowOffset = row * rowBytes;
				updateCpuReadback(surface, m_garbageScratch.data(), rowOffset, rowBytes, 0u, y + static_cast<u32>(row));
			}
			y += static_cast<u32>(rows);
		}
	} else {
		for (u32 y = 0u; y < height; ++y) {
			for (u32 x = 0u; x < surface.surfaceWidth; ) {
				const size_t segmentWidth = std::min<size_t>(maxPixels, surface.surfaceWidth - x);
				const size_t segmentBytes = segmentWidth * 4u;
				fillVramGarbageScratch(m_garbageScratch.data(), segmentBytes, stream);
				updateCpuReadback(surface, m_garbageScratch.data(), 0u, segmentBytes, x, y);
				x += static_cast<u32>(segmentWidth);
			}
		}
	}
}

} // namespace bmsx

File: machine/ts/machine/devices/vdp/vram.ts
import {
	VDP_RD_SURFACE_FRAMEBUFFER,
} from './contracts';
import {
	VRAM_FRAMEBUFFER_BASE,
	VRAM_FRAMEBUFFER_SIZE,
	VRAM_STAGING_BASE,
	VRAM_STAGING_SIZE,
	VRAM_TEXTURE_SIZE,
} from '../../memory/map';
import {
	VRAM_GARBAGE_CHUNK_BYTES,
	VRAM_GARBAGE_SPACE_SALT,
	fillVramGarbageScratch,
	type VramGarbageStream,
} from './vram_garbage';
import type { VdpSurfaceBacking } from './device_output';
import {
	VDP_RPU_PARAM_MEM_PAGE_COUNT,
	bumpVdpRpuVramPageRevisions
} from './rpu';

export type VdpFrameBufferSize = {
	width: number;
	height: number;
};

export type VdpEntropySeeds = {
	machineSeed: number;
	bootSeed: number;
};

export type VdpSurfacePixelsState = {
	surfaceId: number;
	surfaceWidth: number;
	surfaceHeight: number;
	pixels: Uint8Array;
};

export type VdpVramState = {
	rpuVram: Uint8Array;
	surfacePixels: VdpSurfacePixelsState[];
};

function vramSurfaceByteSize(width: number, height: number): number {
	return (width * height * 4) >>> 0;
}

export const DEFAULT_VDP_ENTROPY_SEEDS: VdpEntropySeeds = {
	machineSeed: 0x42564d58,
	bootSeed: 0x7652414d,
};

export class VdpVramUnit {
	private readonly frameBuffer: VdpSurfaceBacking;
	public readonly rpuVram: Uint8Array = new Uint8Array(VRAM_STAGING_SIZE + VRAM_TEXTURE_SIZE);
	public readonly rpuVramPageRevisions: Uint32Array = new Uint32Array(VDP_RPU_PARAM_MEM_PAGE_COUNT);
	private readonly garbageScratch = new Uint8Array(VRAM_GARBAGE_CHUNK_BYTES);
	private machineSeed = DEFAULT_VDP_ENTROPY_SEEDS.machineSeed;
	private bootSeed = DEFAULT_VDP_ENTROPY_SEEDS.bootSeed;

	public constructor(
		private readonly frameBufferSize: VdpFrameBufferSize,
		entropySeeds: VdpEntropySeeds = DEFAULT_VDP_ENTROPY_SEEDS,
	) {
		this.machineSeed = entropySeeds.machineSeed >>> 0;
		this.bootSeed = entropySeeds.bootSeed >>> 0;
		this.frameBuffer = {
			baseAddr: VRAM_FRAMEBUFFER_BASE,
			capacity: VRAM_FRAMEBUFFER_SIZE,
			surfaceId: VDP_RD_SURFACE_FRAMEBUFFER,
			surfaceWidth: frameBufferSize.width,
			surfaceHeight: frameBufferSize.height,
			cpuReadback: new Uint8Array(vramSurfaceByteSize(frameBufferSize.width, frameBufferSize.height)),
		};
		this.seedSurfacePixels(this.frameBuffer);
	}


	public get frameBufferSurface(): VdpSurfaceBacking {
		return this.frameBuffer;
	}

	public initializeFrameBuffer(): void {
		fillVramGarbageScratch(this.rpuVram, this.rpuVram.byteLength, {
			machineSeed: this.machineSeed,
			bootSeed: this.bootSeed,
			slotSalt: VRAM_GARBAGE_SPACE_SALT >>> 0,
			addr: VRAM_STAGING_BASE >>> 0,
		});
		bumpVdpRpuVramPageRevisions(this.rpuVramPageRevisions, 0, this.rpuVram.byteLength);
		this.configureFrameBufferSurface(this.frameBufferSize.width, this.frameBufferSize.height);
	}

	public writeRpuVram(addr: number, bytes: Uint8Array, srcOffset: number, length: number): boolean {
		if (addr < VRAM_STAGING_BASE) {
			return false;
		}
		const offset = addr - VRAM_STAGING_BASE;
		if (offset > this.rpuVram.byteLength || length > this.rpuVram.byteLength - offset) {
			return false;
		}
		for (let index = 0; index < length; index += 1) {
			this.rpuVram[offset + index] = bytes[srcOffset + index]!;
		}
		bumpVdpRpuVramPageRevisions(this.rpuVramPageRevisions, offset, length);
		return true;
	}

	public readRpuVram(addr: number, out: Uint8Array, length: number): boolean {
		if (addr < VRAM_STAGING_BASE) {
			return false;
		}
		const offset = addr - VRAM_STAGING_BASE;
		if (offset > this.rpuVram.byteLength || length > this.rpuVram.byteLength - offset) {
			return false;
		}
		for (let index = 0; index < length; index += 1) {
			out[index] = this.rpuVram[offset + index]!;
		}
		return true;
	}

	public writeSurfaceBytes(surface: VdpSurfaceBacking, offset: number, bytes: Uint8Array, srcOffset: number, length: number): void {
		const stride = surface.surfaceWidth * 4;
		let remaining = length;
		let cursor = srcOffset;
		let row = (offset / stride) >>> 0;
		let rowOffset = offset - row * stride;
		while (remaining > 0) {
			const rowAvailable = stride - rowOffset;
			const rowBytes = remaining < rowAvailable ? remaining : rowAvailable;
			this.updateCpuReadback(surface, bytes, cursor, rowBytes, rowOffset / 4, row);
			remaining -= rowBytes;
			cursor += rowBytes;
			row += 1;
			rowOffset = 0;
		}
	}

	public readSurfaceBytes(surface: VdpSurfaceBacking, offset: number, out: Uint8Array, length: number): void {
		const stride = surface.surfaceWidth * 4;
		let remaining = length;
		let cursor = 0;
		let row = (offset / stride) >>> 0;
		let rowOffset = offset - row * stride;
		const buffer = surface.cpuReadback;
		while (remaining > 0) {
			const rowAvailable = stride - rowOffset;
			const rowBytes = remaining < rowAvailable ? remaining : rowAvailable;
			const srcOffset = row * stride + rowOffset;
			for (let index = 0; index < rowBytes; index += 1) {
				out[cursor + index] = buffer[srcOffset + index]!;
			}
			remaining -= rowBytes;
			cursor += rowBytes;
			row += 1;
			rowOffset = 0;
		}
	}

	public setSurfaceLogicalDimensions(surface: VdpSurfaceBacking, width: number, height: number): void {
		const byteLength = vramSurfaceByteSize(width, height);
		if (surface.surfaceWidth === width && surface.surfaceHeight === height) {
			return;
		}
		const previous = surface.cpuReadback;
		surface.surfaceWidth = width;
		surface.surfaceHeight = height;
		surface.cpuReadback = new Uint8Array(byteLength);
		const copyBytes = previous.byteLength < surface.cpuReadback.byteLength ? previous.byteLength : surface.cpuReadback.byteLength;
		this.seedSurfacePixels(surface);
		for (let index = 0; index < copyBytes; index += 1) {
			surface.cpuReadback[index] = previous[index]!;
		}
	}

	public frameBufferContains(addr: number, length: number): boolean {
		const surface = this.frameBuffer;
		return addr >= surface.baseAddr && addr + length <= surface.baseAddr + surface.capacity;
	}

	public captureState(): VdpVramState {
		const rpuVram = new Uint8Array(this.rpuVram.byteLength);
		for (let index = 0; index < this.rpuVram.byteLength; index += 1) {
			rpuVram[index] = this.rpuVram[index]!;
		}
		return {
			rpuVram,
			surfacePixels: this.captureSurfacePixels(),
		};
	}

	public restoreState(state: VdpVramState): void {
		for (let index = 0; index < state.rpuVram.byteLength; index += 1) {
			this.rpuVram[index] = state.rpuVram[index]!;
		}
		bumpVdpRpuVramPageRevisions(this.rpuVramPageRevisions, 0, state.rpuVram.byteLength);
		for (let index = 0; index < state.surfacePixels.length; index += 1) {
			this.restoreSurfacePixels(state.surfacePixels[index]!);
		}
	}

	public get trackedUsedBytes(): number {
		return this.rpuVram.byteLength + vramSurfaceByteSize(this.frameBuffer.surfaceWidth, this.frameBuffer.surfaceHeight);
	}

	public get trackedTotalBytes(): number {
		return VRAM_STAGING_SIZE + VRAM_TEXTURE_SIZE + VRAM_FRAMEBUFFER_SIZE;
	}

	private configureFrameBufferSurface(width: number, height: number): void {
		const byteLength = vramSurfaceByteSize(width, height);
		const surface = this.frameBuffer;
		surface.baseAddr = VRAM_FRAMEBUFFER_BASE;
		surface.capacity = VRAM_FRAMEBUFFER_SIZE;
		surface.surfaceId = VDP_RD_SURFACE_FRAMEBUFFER;
		surface.surfaceWidth = width;
		surface.surfaceHeight = height;
		surface.cpuReadback = new Uint8Array(byteLength);
		this.seedSurfacePixels(surface);
	}

	private captureSurfacePixels(): VdpSurfacePixelsState[] {
		const surface = this.frameBuffer;
		const pixels = new Uint8Array(surface.cpuReadback.byteLength);
		for (let byteIndex = 0; byteIndex < surface.cpuReadback.byteLength; byteIndex += 1) {
			pixels[byteIndex] = surface.cpuReadback[byteIndex]!;
		}
		return [{
			surfaceId: surface.surfaceId,
			surfaceWidth: surface.surfaceWidth,
			surfaceHeight: surface.surfaceHeight,
			pixels,
		}];
	}

	private restoreSurfacePixels(state: VdpSurfacePixelsState): void {
		const surface = this.frameBuffer;
		for (let index = 0; index < state.pixels.byteLength; index += 1) {
			surface.cpuReadback[index] = state.pixels[index]!;
		}
	}

	private updateCpuReadback(surface: VdpSurfaceBacking, bytes: Uint8Array, srcOffset: number, length: number, x: number, y: number): void {
		const buffer = surface.cpuReadback;
		const stride = surface.surfaceWidth * 4;
		const offset = y * stride + x * 4;
		for (let index = 0; index < length; index += 1) {
			buffer[offset + index] = bytes[srcOffset + index]!;
		}
	}

	private seedSurfacePixels(surface: VdpSurfaceBacking): void {
		const width = surface.surfaceWidth;
		const height = surface.surfaceHeight;
		const rowPixels = width;
		const maxPixels = this.garbageScratch.byteLength >>> 2;
		const stream: VramGarbageStream = {
			machineSeed: this.machineSeed,
			bootSeed: this.bootSeed,
			slotSalt: VRAM_GARBAGE_SPACE_SALT >>> 0,
			addr: surface.baseAddr >>> 0,
		};
		if (rowPixels <= maxPixels) {
			const rowsPerChunk = (maxPixels / rowPixels) >>> 0;
			for (let y = 0; y < height;) {
				const rowsRemaining = height - y;
				const rows = rowsPerChunk < rowsRemaining ? rowsPerChunk : rowsRemaining;
				const chunkBytes = rowPixels * rows * 4;
				fillVramGarbageScratch(this.garbageScratch, chunkBytes, stream);
				for (let row = 0; row < rows; row += 1) {
					const rowOffset = row * rowPixels * 4;
					this.updateCpuReadback(surface, this.garbageScratch, rowOffset, rowPixels * 4, 0, y + row);
				}
				y += rows;
			}
		} else {
			for (let y = 0; y < height; y += 1) {
				for (let x = 0; x < width;) {
					const widthRemaining = width - x;
					const segmentWidth = maxPixels < widthRemaining ? maxPixels : widthRemaining;
					const segmentBytes = segmentWidth * 4;
					fillVramGarbageScratch(this.garbageScratch, segmentBytes, stream);
					this.updateCpuReadback(surface, this.garbageScratch, 0, segmentBytes, x, y);
					x += segmentWidth;
				}
			}
		}
	}
}
